File: agent-go/internal/tools/find.go
package tools

import (
	"context"
	"io/fs"
	"path/filepath"
	"strings"
)

type FindTool struct{ cwd string }

func NewFindTool(cwd string) *FindTool { return &FindTool{cwd: cwd} }

func (t *FindTool) Name() string        { return "find" }
func (t *FindTool) Description() string { return "Recursively find files or directories matching a glob pattern." }
func (t *FindTool) Schema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"path":    map[string]any{"type": "string", "description": "Directory to search in"},
			"pattern": map[string]any{"type": "string", "description": "Filename glob pattern (e.g. *.go)"},
			"type": map[string]any{
				"type":        "string",
				"enum":        []string{"file", "dir", "any"},
				"description": "Entry type filter (default: any)",
			},
		},
		"required": []string{"path", "pattern"},
	}
}

func (t *FindTool) Execute(_ context.Context, _ string, args map[string]any, _ UpdateFunc) (ToolResult, error) {
	root, _ := args["path"].(string)
	pattern, _ := args["pattern"].(string)
	typeFilter, _ := args["type"].(string)
	if root == "" {
		return errResult("find: missing required parameter 'path'"), nil
	}
	if pattern == "" {
		return errResult("find: missing required parameter 'pattern'"), nil
	}
	root = resolvePath(root, t.cwd)

	var results []string
	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // skip unreadable
		}
		if path == root {
			return nil
		}
		matched, _ := filepath.Match(pattern, d.Name())
		if !matched {
			return nil
		}
		switch typeFilter {
		case "file":
			if d.IsDir() {
				return nil
			}
		case "dir":
			if !d.IsDir() {
				return nil
			}
		}
		rel, _ := filepath.Rel(root, path)
		results = append(results, rel)
		return nil
	})
	if err != nil {
		return errResult("find: " + err.Error()), nil
	}
	if len(results) == 0 {
		return ToolResult{Content: "(no matches)"}, nil
	}
	return ToolResult{Content: strings.Join(results, "\n")}, nil
}

File: agent-rust/src/tools/find.rs
use async_trait::async_trait;
use tokio_util::sync::CancellationToken;
use walkdir::WalkDir;

use crate::tools::{Tool, ToolResult, UpdateFn};

pub struct FindTool;

#[async_trait]
impl Tool for FindTool {
    fn name(&self) -> &str { "find" }
    fn description(&self) -> &str {
        "Find files or directories matching a glob pattern under a path."
    }
    fn schema(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "path":    {"type": "string", "description": "Directory to search"},
                "pattern": {"type": "string", "description": "Filename glob (e.g. '*.rs')"},
                "type":    {
                    "type": "string",
                    "enum": ["file", "dir", "any"],
                    "description": "Filter by entry type (default: any)"
                }
            },
            "required": ["path", "pattern"]
        })
    }

    async fn execute(
        &self,
        _id: &str,
        args: serde_json::Value,
        _cancel: CancellationToken,
        _on_update: Option<&UpdateFn>,
    ) -> ToolResult {
        let root = args["path"].as_str().unwrap_or(".").to_string();
        let pattern_str = match args["pattern"].as_str() {
            Some(p) => p.to_string(),
            None    => return ToolResult::err("missing 'pattern'"),
        };
        let filter_type = args["type"].as_str().unwrap_or("any").to_string();

        let glob_pat = match glob::Pattern::new(&pattern_str) {
            Ok(p)  => p,
            Err(e) => return ToolResult::err(format!("invalid glob pattern: {e}")),
        };

        let mut matches: Vec<String> = Vec::new();

        for entry in WalkDir::new(&root).into_iter().filter_map(|e| e.ok()) {
            let is_dir = entry.file_type().is_dir();
            let matches_type = match filter_type.as_str() {
                "file" => !is_dir,
                "dir"  => is_dir,
                _      => true,
            };

            let file_name = entry.file_name().to_str().unwrap_or("");
            if matches_type && glob_pat.matches(file_name) {
                // Return path relative to root
                let rel = entry.path()
                    .strip_prefix(&root)
                    .unwrap_or(entry.path())
                    .display()
                    .to_string();
                if !rel.is_empty() {
                    matches.push(rel);
                }
            }
        }

        matches.sort();

        if matches.is_empty() {
            ToolResult::ok("(no matches)")
        } else {
            ToolResult::ok(matches.join("\n"))
        }
    }
}
